src/Entity/BannerSelection/Campaign.php

```diff
@@ -12,24 +14,37 @@ class Campaign {
 	private $identifier;
 	private $start;
 	private $end;
+	private $rng;
+	private $displayRatio;
 
 	/**
 	 * @var Bucket[]
 	 */
 	private $buckets;
 
-	public function __construct( string $identifier, \DateTime $start, \DateTime $end, array $buckets ) {
+	public function __construct(
+		string $identifier,
+		\DateTime $start,
+		\DateTime $end,
+		float $displayRatio,
+		RandomIntegerInterface $rng,
+		Bucket $mainBucket,
+		Bucket ...$additionalBuckets ) {
```

Comment:
There might be campaigns where there is no A/B test, so I guess just a variadic `$buckets` argument without `$mainBucket` would be better.

Reply:
No, you are wrong, Gabriel! Very wrong, sad!